services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceControllerStateModel.java

```diff
@@ -101,23 +114,49 @@ public boolean updateState(String newState) {
     return result;
   }
 
-  private void executeStateTransition(Message message, StateTransition stateTransition) throws VeniceException {
-    String from = message.getFromState();
-    String to = message.getToState();
-    String threadName = "Helix-ST-" + message.getResourceName() + "-" + from + "->" + to;
-    // Change name to indicate which st is occupied this thread.
-    Thread.currentThread().setName(threadName);
+  /**
+   * Executes the state transition synchronously with a thread name prefix "Sync-Helix-ST".
+   */
+  private void executeStateTransitionSync(Message message, StateTransition stateTransition) {
+    String threadName = String
+        .format("Sync-Helix-ST-%s-%s->%s", message.getResourceName(), message.getFromState(), message.getToState());
+    executeStateTransitionWithThreadName(threadName, stateTransition);
+  }
+
+  /**
+   * Executes the state transition asynchronously with a thread name prefix "Async-ClusterName-Helix-ST".
+   */
+  Future<?> executeStateTransitionAsync(Message message, StateTransition stateTransition) {
+    String threadName = String.format(
+        "Async-%s-Helix-ST-%s-%s->%s",
+        clusterName,
+        message.getResourceName(),
+        message.getFromState(),
+        message.getToState());
+    return workerService.submit(() -> {
+      executeStateTransitionWithThreadName(threadName, stateTransition);
+    });
+  }
+
+  /**
+   * Core method that runs the state transition with a custom thread name.
+   * The thread name is set for debugging purposes.
+   */
+  private void executeStateTransitionWithThreadName(String threadName, StateTransition stateTransition) {
+    Thread currentThread = Thread.currentThread();
+    String originalName = currentThread.getName();
+    currentThread.setName(threadName);
     try {
       stateTransition.execute();
     } catch (Exception e) {
       throw new VeniceException("Failed to execute '" + threadName + "'.", e);
```

Comment:
Correct me if I'm wrong but I think in the `executeStateTransitionAsync` code path the exception we catch and throw here will simply be swallowed since we just call `workerService.submit` without ever getting/checking the returned future. Let's log the exception that we catch here?

Reply:
I think this is a good point! Let me add a bit more details regarding your points.

For `standby-to-leader`, if the inner `Controller-ST-Worker` thread throws any exception and because we call `Future.get` to wait its result, catch exceptions from worker thread, and fail Helix ST, we maintain almost the same semantics as of today (except introducing a timeout).

However, for `leader-to-standby`, this PR will change it to aync way, exceptions from `executeStateTransitionAsync` will not be caught by the original ST thread. Any exceptions, that could fail ST previously, will no longer do so.  Because of that we should log any issues happened in the inner thread, otherwise it will be a silent failure.